### DISCUSSION TITLE
Why you need your own parser and don't use bundler esbuild, rollup, ...

### DISCUSSION THREAD

**Discussion by parweb** [phero-hq/phero#75] [Ideas]
Hi

When I first try phero
I immediately wanted to organise my services into many sub files.

for me it was
```
- user
---- index.ts
---- list.ts
---- User.ts
- shared
---- index.ts
---- Id.ts
- post
---- index.ts
---- list.ts
---- Post.ts
```

then of course import each other when needed

```ts
import { User } from '../user';

const collection: User[] = [
  { id: 1, name: 'chris', posts: [1, 2] },
  { id: 2, name: 'cassie', posts: [3, 4] },
  { id: 3, name: 'lena', posts: [5, 6] },
  { id: 4, name: 'brahim', posts: [7, 8] }
];

export default function list(limit: number): User[] {
  return collection;
}
```

But I ran into limitation (for now) by your parser implementation.

So I made a few researches on bundler
And I manage to make a P.O.C. with tsc & dts-bundle

https://github.com/parweb/phero-export-imported-types/

the main point is this in ```packages.json``` -> scripts ```export```
```tsc --emitDeclarationOnly --outDir dist && dts-bundle --name=generated --main=dist/sample/index.d.ts```
then see ```./dist/sample/generated.d.ts```

I also try to export imported types from node_modules
like prisma

but my knowledge on bundler fail short :-)
to automatically include types from external library without the need to specify which packages I want in ```tsconfig.json```

I hope it will start a public debate on this subject ;-)

**Comment by JasperH8g** [comment]
First of all, thanks for reaching out and more importantly: trying out Phero! 🥳 

To make me understand your problem a bit better, can you elaborate on what you're trying to do?

With Phero, it shouldn't be a problem to organise your types and functions in any way that you'd like. One very important thing about this though is that a Phero-server-project should always start with a file at `server-project/src/phero.ts` (the "Phero-file"). This is the point where you can export your services and Phero should be able to figure out what types/functions you need on the frontend and ship those in the generated client. 

Small sidetone on this: We're working on a feature that makes it possible to export any type from your Phero-file and use it in your client, even if it isn't used in any of your services/functions, but we aren't there yet. For now: **All types that are used by your functions should end up in the client**.

So, before going into extra tools like dts-bundle or anything like that, I'd like to know: How does your Phero-file looks like, and what external type specifically is causing parser-limitations?

**Comment by parweb** [comment]
My bad I just realize that I was doing export default on every file and when changed it to export only everything works ^^

**Comment by JasperH8g** [comment]
Ah great! Let us know if we can help with anything else 😊
